### DISCUSSION TITLE
Compliance tensor is equal to nan

### DISCUSSION THREAD

**Discussion by RossMech** [idaholab/moose#21189] [Q&A Modules: Solid mechanics] [answered]
Hello everyone,

I have following problem. I would like to calculate the phase-field stress distribution using Reuss-Sachs approximation, i.e.

$\mathbf{C} = \left( \sum h_{\alpha} \mathbf{S}_{\alpha} \right)^{-1}$ 
 
where $ h_{\alpha} $ is weighting function of the each particular phase and $\mathbf{S}_{\alpha}$ is compliance tensor of each phase.

My idea was to calculate the compliance tensor of each phase hard-coded inside of the stress calculation subroutine, i.e.
`RankFourTensor compliance_alpha = _elasticity_tensor_alpha[_qp].inverse();`

The elasticity tensor is constructed with ComputeIsotropicElasticityTensor material, i.e.
  [./elasticity_tensor_alpha]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1
    poissons_ratio = 0.3
    base_name = alpha_phase
  [../]
  
The output inverse is estimated to be nan. If someone has idea, why the inversion operation results in nan values, I would like to discuss them!

Best regards,
Rostyslav Nizinkovskyi.

**Comment by hugary1995** [comment, marked as answer]
It shouldn't be nan given this pair of lame parameters. Where and how are you calculating the `_elasticity_tensor_alpha`?

**Reply by hugary1995** [reply]
For example, if you calculate the inverse before elasticity tensor is computed, you'll for sure get nan. You could try printing out the elasticity tensor right before the inversion.

**Reply by RossMech** [reply]
Hi, 

thank you for the reply and suggestions. `_elasticity_tensor_alpha` is consumed into material subroutine via

`params.addRequiredParam<std::string>("base_name_alpha","Elasticity tensor of alpha phase eta = 1");`

and

`_base_name_alpha(getParam<std::string>("base_name_alpha") + "_"),
    _elasticity_tensor_alpha(getMaterialPropertyByName<RankFourTensor>(_base_name_alpha+"elasticity_tensor")),`.

I had also tried to get output for the all values, which should be non-zero

![image](https://user-images.githubusercontent.com/62218067/171824105-919c1d6b-c60c-4bb3-aedd-c5793a56e665.png)

As you can see, the values are non-zero and have a correspondent values. But the inversion is nevertheless nan (except for one value). 

If you have further suggestions, I would like to consider them as well!

**Reply by hugary1995** [reply]
Okay, I can reproduce this. I will create an issue out of this discussion.

For now, since your elasticity tensor is symmetric, you can use `invSymm` in place of `inverse` as a work-around.

@dschwen FYI. This bug is related to eigen I believe.

**Reply by RossMech** [reply]
Hi! The invSymm method seems to work just correct! Thank you very much for the answer.

**Comment by ngrilli** [comment]
Dear @RossMech

You can see an example of inversion of RankFourTensor here:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/unit/src/RankFourTensorTest.C

You need to specify where are you carrying out the inversion.
Is that inside a material object?

In that case, it is better to avoid doing the inversion inside the initQpStatefulProperties() function
because at that stage the elasticity tensor may not be yet calculated. 

Best Regards,
Nicolò

**Reply by RossMech** [reply]
Thank you for the answer! I use a ComputeStressBase as a starting point for my material development. The inversion is calculated inside of **ComputeQpStress()** function.

**Reply by RossMech** [reply]
Also a little bit of additional information. In answer to @hugary1995 I had added an image, containing output of non-zero values of the stiffness and compliance tensors.

**Reply by ngrilli** [reply]
@RossMech  difficult to say, either the elasticity is not invertible for some reason, like components don't get filled completely. Or the class you use cannot be inverted. I suggest the following attempts to clarify:
1) Use without base name
2) Try printing _elasticity_tensor_alpha after the inversion operation, maybe inverse() is not meant to give an output and just invert the tensor
3) Try using another class instead of ComputeIsotropicElasticityTensor, like ComputeElasticityTensorConstantRotationCP, like here:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/linesearch.i
and try inverting _Cijkl instead.
If you can share the entire material class, it is easier to help.
Nicolò

**Reply by hugary1995** [reply]
1 & 3 aren't related to this issue as I see it.
2. `inverse()` isn't an in-place operation.

**Reply by ngrilli** [reply]
@RossMech  do you include with 
#include "RankFourTensor.h"
in your material class?

**Reply by hugary1995** [reply]
@ngrilli I am discussing this in #21211, feel free to take a look if you are interested.
